### PR TITLE
Support 'preload' attribute for Audio Block

### LINKED ISSUE #6581: Use of loop and autostart in <audio> causes unknown error
## Issue Overview

Use of loop and autostart in <audio> causes unknown error

## Steps to Reproduce (for bugs)

1. Add audio block
2. Use edit as HTML to add the loop and/or autoplay attributes to the<audio> tag.
3. Update the post.
4. message: block has been changed
5. Click edit as HTML in the block.
6. block shows unknown error. No HTML is shown
7. Switch to code editor: the string ="" is added to the controls, loop and autoplay attributes

Firefox 59.0.3, Chrome 66, Gutenberg 2.8

## Expected Behavior

Preferably the addition of loop and autoplay options(checkboxes), as are available when using the audio shortcode.
At least allow for the error free addition of attributes.

### PR DESCRIPTION
**Pull request by danielbachhuber**
## Description

Adds UI for managing `preload` attribute of the Audio Block:

image

This is for feature parity with the `[audio]` shortcode:

image

The default value is 'None'. Selecting 'Auto' or 'Metadata' will add `preload="auto"` or `preload="meta"` to the `<audio>` element, respectively.

Fixes #6581
Previously #7322

## How has this been tested?

1. Add an Audio Block to the editor. View that it has no `preload` attribute set by default:

image

image

2. Switch back to Visual mode, select "Auto", and then view that `preload='auto'` is set:

image

image

3. Save the post and verify the `<audio>` element includes `preload="auto"`:

image

## Types of changes

Enhancement to existing feature.

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation.

### CODE REVIEW

File: core-blocks/audio/edit.js
Comment (danielbachhuber) [inline core-blocks/audio/edit.js]:
```
attributeValue: undefined,
```
Worth noting I was stuck on this for a little bit. Originally, I tried `attributeValue: false` but ended up with `preload="false"` in the saved `<audio>` attribute.

**Comment by danielbachhuber** [comment]
Flagged `Needs Design Feedback` to get approval on this UI:

image

**Review by karmatosed — DISMISSED** [review]
I think we need to think about the actual UI here. I'll add a note to make a mockup here but let's not merge this as currently we don't use this tab/button combination anywhere and ideally we'd not be adding it. I know WP core does but it's not super effective as a pattern.

One thought though before mocking up, why not a drop down? Can you select more than one or just one? Also what do these settings mean to everyone? I ask as exposing them in the UI like this for the block seems like it needs an explanation.

**Comment by danielbachhuber** [comment]
> I'll add a note to make a mockup here but let's not merge this as currently we don't use this tab/button combination anywhere and ideally we'd not be adding it.

Ok, I've moved to the 3.3 milestone.

> One thought though before mocking up, why not a drop down? Can you select more than one or just one?

Personally, I prefer the current UI because I can see all potential options. You can only select one option. It's similar to the existing Text Settings UI:

image

> I ask as exposing them in the UI like this for the block seems like it needs an explanation.

It's for feature-parity with the Classic Editor.

**Comment by karmatosed** [comment]
Let's leave it in with tab button UI but also let's add a description as it's kind of not as intuitive what this means compared to text settings.

**Comment by karmatosed** [comment]
Thinking about this a bit more let's combine to one group, no sections and stay with drop down it does make far more sense here.

2018-07-05 at 12 00

Also let's have a description if that makes sense or a smaller one can be thought of.

File: core-blocks/audio/edit.js
Comment (tofumatt) [inline core-blocks/audio/edit.js]: I know "auto" is the HTML value for this, but we don't have to follow the HTML attributes to a T.

I think a more instructive/helpful label here would be "Content" (as that's what you're instructing the browser to preload) or, if you want more text: "Preload Content". If you really want the HTML value in brackets you could do "Content (Auto)" but I think "Content would suffice and then nix the need for a description beneath.

File: core-blocks/audio/edit.js
Comment (tofumatt) [inline core-blocks/audio/edit.js]: "Metadata only" might be a better label especially if my other suggestion was implemented.

**Review by tofumatt — COMMENTED** [review]
I have a thought about how to label these without needing more description text. I think the issue is we're just being too close to HTML attribute with these labels.

File: core-blocks/audio/edit.js
Comment (tofumatt) [reply core-blocks/audio/edit.js]: If `null` works I might think that is ever-so-slightly more semantically correct? If it worked I think it might be nicer, but yeah this is fine 😄

File: core-blocks/audio/index.js
Comment (tofumatt) [inline core-blocks/audio/index.js]: This will need a deprecated `save` function, right? (https://github.com/WordPress/gutenberg/blob/master/docs/block-api/deprecated-blocks.md)

Or will this load fine because it's just adding the binary attrs?

File: core-blocks/audio/edit.js
Comment (danielbachhuber) [reply core-blocks/audio/edit.js]: `null` didn't work in my original testing.

File: core-blocks/audio/index.js
Comment (danielbachhuber) [reply core-blocks/audio/index.js]: The latter, to my knowledge. The markup doesn't change structure; this is only a progressive enhancement.

Comment (tofumatt) [reply core-blocks/audio/index.js]: Oh right, cool. 👍 (Sorry I'm still a bit new to block validation!)

**Comment by danielbachhuber** [comment]
> Thinking about this a bit more let's combine to one group, no sections and stay with drop down it does make far more sense here.

@karmatosed My `<SelectControl>` ends up looking like this:

image

How do I make it look like yours?

File: core-blocks/audio/edit.js
Comment (tofumatt) [inline core-blocks/audio/edit.js]: Cheers! 👍

File: core-blocks/audio/edit.js
Comment (danielbachhuber) [reply core-blocks/audio/edit.js]: I'm not particularly keen to bikeshed the labels at this point.

If you'd like to take the changes through the copy review process, then please do.

**Comment by danielbachhuber** [comment]
@karmatosed @tofumatt Could I get clear and specific feedback on how you'd like me to finish this pull request?

**Review by tofumatt — CHANGES_REQUESTED** [review]
Tested locally, including going from a post with existing audio to this one and it all worked for me.

The code makes sense. The removal of `PanelBody` makes sense to me but it means the "Preload" label is really tight against the `Audio` block description:

screenshot 2018-07-05 15 18 06

It might make sense to bring it back. Compare it to:

screenshot 2018-07-05 15 22 57

Which I think is nicer. I think that's what @karmatosed is talking about here.

-----

I'd be fine putting "preload" in advanced controls, frankly. Let's just do that–it's not a playback control so restoring the "Playback controls" to how they were before and moving "Preload" under *Advanced* makes sense to me.

**Comment by danielbachhuber** [comment]
> t might make sense to bring it back. Compare it to:
> 
> Which I think is nicer. I think that's what @karmatosed is talking about here.

@karmatosed Are you 👍 on @tofumatt's proposed implementation?

**Comment by karmatosed** [comment]
All blocks should really have that header block as @tofumatt points out. I also think we can change the order around a bit. What about this?

artboard

As to if this is an advanced setting or not.. I am unsure. Let's have what I suggest as a middle ground and iterate from there once the PR is in.

**Comment by danielbachhuber** [comment]
> I also think we can change the order around a bit. What about this?

Works for me. The block now looks like this with eed2f1b15:

image

**Review by tofumatt — APPROVED** [review]

**Comment by tofumatt** [comment]
I made a little tweak (d1f70622fcbc02b071e33e008e321c4d6b0aceed) because I found the constant-first ternary check was tough to follow. I think we should eschew constant-first styles especially as we're inconsistent about them, but that's not for this issue 😆

If you're okay with that, then 🚢, looks great!

**Pull request merged**
